MENTAL-MODEL.md

```diff
@@ -0,0 +1,376 @@
+
+fn try_contributing_inputs(
+    payjoin: &mut PayjoinProposal,
+    bitcoind: &bitcoincore_rpc::Client,
+) -> Result<()> {
+    use bitcoin::OutPoint;
+
+    let available_inputs = bitcoind
+        .list_unspent(None, None, None, None, None)
+        .context("Failed to list unspent from bitcoind")?;
+    let candidate_inputs: HashMap<Amount, OutPoint> = available_inputs
+        .iter()
+        .map(|i| (i.amount, OutPoint { txid: i.txid, vout: i.vout }))
+        .collect();
+
+    let selected_outpoint = payjoin.try_preserving_privacy(candidate_inputs).expect("gg");
+    let selected_utxo = available_inputs
+        .iter()
+        .find(|i| i.txid == selected_outpoint.txid && i.vout == selected_outpoint.vout)
+        .context("This shouldn't happen. Failed to retrieve the privacy preserving utxo from those we provided to the seclector.")?;
+    log::debug!("selected utxo: {:#?}", selected_utxo);
+
+    //  calculate receiver payjoin outputs given receiver payjoin inputs and original_psbt,
+    let txo_to_contribute = bitcoin::TxOut {
+        value: selected_utxo.amount.to_sat(),
+        script_pubkey: selected_utxo.script_pub_key.clone(),
+    };
+    let outpoint_to_contribute =
+        bitcoin::OutPoint { txid: selected_utxo.txid, vout: selected_utxo.vout };
+    payjoin.contribute_witness_input(txo_to_contribute, outpoint_to_contribute);
+    Ok(())
+}
+```
+
+Serious, in-depth research has gone into proper transaction construction. [Here's a good starting point from the JoinMarket repo](https://gist.github.com/AdamISZ/4551b947789d3216bacfcb7af25e029e#gistcomment-2796539). Using methods for coin selection not provided by this library may have dire implications for privacy.
+
+### 6. Extract the payjoin PSBT and sign it
+
+Fees are applied to the augmented Payjoin Proposal PSBT using calculation factoring the receiver's own preferred feerate and the sender's fee-related [optional parameters](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#optional-parameters). The current `apply_fee` method is primitive, disregarding PSBT fee estimation and only adding fees coming from the sender's budget. When more accurate tools are available to calculate a PSBT's fee-dependent weight (slightly more complicated than it sounds, but solved, just unimplemented in rust-bitcoin), this `apply_fee` should be improved.
+
+```rs
+let payjoin_proposal_psbt = payjoin.apply_fee(min_feerate_sat_per_vb: Some(1))?;
+
+log::debug!("Extracted PSBT: {:#?}", payjoin_proposal_psbt);
+// Sign payjoin psbt
+let payjoin_base64_string =
+    base64::encode(bitcoin::consensus::serialize(&payjoin_proposal_psbt));
+// `wallet_process_psbt` adds available utxo data and finalizes
+let payjoin_proposal_psbt =
+    bitcoind.wallet_process_psbt(&payjoin_base64_string, sign: None, sighash_type: None, bip32derivs: Some(false))?.psbt;
+let payjoin_proposal_psbt =
+    load_psbt_from_base64(payjoin_proposal_psbt.as_bytes()).context("Failed to parse PSBT")?;
+```
```

Comment:
The signing algorithm could be passed as a closure, but may encounter runtime issues as mentioned previously. I.e., rust does not yet support async closures.

Reply:
While bitcoind's `wallet_process_psbt` handles most everything you throw at it with grace, LND's PSBT signing [gets very cranky](https://github.com/lightningnetwork/lnd/issues/7266) when fed inputs belonging to other wallets and requires [a complicated workaround of multiple gRPC calls](https://github.com/chaincase-app/nolooking/blob/a2be2adbf2f75074e9e0d913f761e834b8ce70d1/src/scheduler.rs#L367-L410) to function properly. It might be good to note this in the documentation since it's such a pain in the kiester.